cores/esp32/esp32-hal-rmt.c

```diff
@@ -13,622 +13,556 @@
+  RMT_MUTEX_UNLOCK(bus);
+  return retCode;
 }
 
-rmt_obj_t* rmtInit(int pin, bool tx_not_rx, rmt_reserve_memsize_t memsize)
+bool rmtInit(int pin, rmt_ch_dir_t channel_direction, rmt_reserve_memsize_t mem_size, uint32_t frequency_Hz)
 {
-    uint32_t buffers = memsize;
-    rmt_obj_t* rmt = NULL;
-    uint32_t i = 0;
-    uint32_t j = 0;
-
-    // create common block mutex for protecting allocs from multiple threads
-    if (!g_rmt_block_lock) {
-        g_rmt_block_lock = xSemaphoreCreateMutex();
-    }
-    // lock
-    while (xSemaphoreTake(g_rmt_block_lock, portMAX_DELAY) != pdPASS) {}
-
-    // Some SoC may have fixed channel numbers for TX and RX - example: ESP32C3
-    uint8_t ch_start, ch_end;
-    if (tx_not_rx) {
-        ch_start = RMT_TX_CH_START;
-        ch_end = RMT_TX_CH_END;
-    } else {
-        ch_start = RMT_RX_CH_START;
-        ch_end = RMT_RX_CH_END;
-    }
-    for (i=ch_start; i<=ch_end; i++) {
-        for (j=0; j<buffers && i+j <= ch_end; j++) {
-            // if the space is ocupied break and continue on other channel
-            if (g_rmt_objects[i+j].allocated) {
-                i += j; // continue searching from latter channel
-                break;
-            }
-        }
-        if (j == buffers) {
-            // found a space in channel descriptors
-            break;
-        }
-    }
-    if (i == MAX_CHANNELS || i+j > MAX_CHANNELS || j != buffers)  {
-        xSemaphoreGive(g_rmt_block_lock);
-        log_e("rmInit Failed - not enough channels");
-        return NULL;
+  log_v("GPIO %d - %s - MemSize[%d] - Freq=%dHz", pin, channel_direction == RMT_RX_MODE ? "RX MODE" : "TX MODE", mem_size * RMT_SYMBOLS_PER_CHANNEL_BLOCK, frequency_Hz);
+
+  // create common block mutex for protecting allocs from multiple threads allocating RMT channels
+  if (!g_rmt_block_lock) {
+    g_rmt_block_lock = xSemaphoreCreateMutex();
+    if (g_rmt_block_lock == NULL) {
+      log_e("GPIO %d - Failed creating RMT Mutex.", pin);
+      return false;
+    }
+  }
+
+  // set Peripheral Manager deInit Callback
+  perimanSetBusDeinit(ESP32_BUS_TYPE_RMT_TX, _rmtDetachBus);
+  perimanSetBusDeinit(ESP32_BUS_TYPE_RMT_RX, _rmtDetachBus);
+
+  // check is pin is valid and in the right direction
+  if ((channel_direction == RMT_TX_MODE && !GPIO_IS_VALID_OUTPUT_GPIO(pin)) || (!GPIO_IS_VALID_GPIO(pin))) {
+    log_e("GPIO %d is not valid or can't be used for output in TX mode.", pin);
+    return false;
+  }
+
+  // validate the RMT ticks by the requested frequency
+  // Based on 80Mhz using a divider of 8 bits (calculated as 1..256)
+  if (frequency_Hz > 80000000 || frequency_Hz < 312500) {
+    log_e("GPIO %d - Bad RMT frequency resolution. Must be between 312.5KHz to 80MHz.", pin);
+    return false;
+  }
+
+  // Try to dettach any (Tx|Rx|Whatever) previous bus or just keep it as not attached
+  if (perimanGetPinBusType(pin) != ESP32_BUS_TYPE_INIT && !perimanSetPinBus(pin, ESP32_BUS_TYPE_INIT, NULL)) {
```

Comment:
What if `perimanGetPinBusType(pin)` will return `ESP32_BUS_TYPE_INIT`, then the IF will fail. I think that part of checking `perimanGetPinBusType(pin) != ESP32_BUS_TYPE_INIT` can be removed and just leave the second part of IF.